ckan/plugins/blanket.py

```diff
@@ -0,0 +1,251 @@
+
+In addition, if plugin follows custom naming conventions, it's
+possible to customize implementation, by providing argument to
+decorator.
+
+If extension uses different names for modules::
+
+    import ckanext.ext.custom_actions as custom_module
+
+    @tk.blanket.action(custom_module)
+    class MyPlugin(plugins.SingletonPlugin):
+        pass
+
+If extension already defines function that returns items required by
+interface::
+
+    def all_actions():
+        return {'ext_action': ext_action}
+
+    @tk.blanket.action(all_actions)
+    class MyPlugin(plugins.SingletonPlugin):
+        pass
+
+If extension statically defines collection of items required by
+interface::
+
+    all_actions = {'ext_action': ext_action}
+
+    @tk.blanket.action(all_actions)
+    class MyPlugin(plugins.SingletonPlugin):
+        pass
+
+"""
+import logging
+import enum
+import types
+
+from collections import namedtuple
+from functools import update_wrapper
+from importlib import import_module
+
+import ckan.plugins as p
+
+__all__ = [u"helper", u"auth", u"action", u"blueprint", u"cli", u"validator"]
+
+log = logging.getLogger(__name__)
+
+
+class Blanket(enum.IntEnum):
+    """Enumeration of all available blanket types.
+
+    In addition, contains hidden `_all` option, that contains all
+    other types. This option is experimental and shouldn't be used
+    outside current module, as it can be removed in future.
+
+    """
+
+    helper = 1 << 0
+    auth = 1 << 1
+    action = 1 << 2
+    blueprint = 1 << 3
+    cli = 1 << 4
+    validator = 1 << 5
+    _all = helper | auth | action | blueprint | cli | validator
+
+    def path(self):
+        """Return relative(start from `ckanext.ext`) import path for
+        implementation.
+
+        """
+        return _mapping[self].path
+
+    def method(self):
+        """Return the name of the method, required for implementation.
+
+        """
+        return _mapping[self].method
+
+    def interface(self):
+        """Return interface provided by blanket.
+        """
+        return _mapping[self].interface
+
+    def returns_list(self):
+        """Check, whether implementation returns list instead of dict.
+
+        """
+        return self & (Blanket.cli | Blanket.blueprint)
+
+    def implement(self, locals, plugin, subject):
+        """Provide implementation for interface.
+
+        """
+        if subject is None:
+            _last_dot = plugin.__module__.rindex(u".")
+            root = plugin.__module__[:_last_dot]
+            import_path = u".".join([root, self.path()])
+            try:
+                subject = import_module(import_path)
+            except ImportError:
+                log.debug(
+                    u"Unable to import <%s> for "
+                    u"blanket implementation of %s for %s",
+                    import_path,
+                    self.interface().__name__,
+                    plugin.__name__,
+                )
+                return
```

Comment:
i think I'd rather not hide this exception, it means the plugin is broken, right?

Reply:
Yes. I think I've used this catch initially block for `blanket.all` which should implement all available interfaces. And then we decided that `blanket.all` is too unpredictable, but I forgot about this block